src/main/java/org/opentripplanner/apis/gtfs/PatternByServiceDatesFilter.java

```diff
@@ -0,0 +1,109 @@
+package org.opentripplanner.apis.gtfs;
+
+import java.time.LocalDate;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.opentripplanner.apis.gtfs.generated.GraphQLTypes.GraphQLServiceDateFilterInput;
+import org.opentripplanner.transit.model.network.Route;
+import org.opentripplanner.transit.model.network.TripPattern;
+import org.opentripplanner.transit.model.timetable.Trip;
+import org.opentripplanner.transit.service.TransitService;
+
+/**
+ * Encapsulates the logic to filter patterns by the service dates that they operate on. It also
+ * has a method to filter routes by checking if their patterns operate on the required days
+ */
+public class PatternByServiceDatesFilter {
```

Comment:
This class implements business logic that should be generelized and shared. This is probably not the only example of this, so we might wait to push this into the transit-domain until we have refactored it. This increases the technical dept.

Reply:
Also, I do not just want to push this into the TransitService, I want us to have a flexible filtering capability on the transit service witch can be reused across all use-cases which needs it. We should be able to filter on more than just serviceDates (runningDates, mode +++) and return other entities than just patterns - eg agencies operating a stop in a given period.

Reply:
Ok, I understand the goal but what are the actionable tasks for this PR?

Reply:
I added a coment in https://github.com/opentripplanner/OpenTripPlanner/pull/5869/commits/f549d053c39a6dbb84fa94b50f74b851fa842d65